frontend/ui/menusorter.lua

```diff
@@ -0,0 +1,151 @@
+            for order_number,order_number_id in ipairs(order_item) do
+                DEBUG("order_number,order_number_id", order_number,order_number_id)
+
+                -- this is a submenu, mark it for later
+                if order[order_number_id] then
+                    table.insert(self.sub_menus, order_number_id)
+                    self.menu_table[order_id][order_number] = {
+                        id = order_number_id,
+                        --sub = true,
+                    }
+                -- regular, just insert a menu action
+                else
+                    --self.menu_table[order_id] = tmp_menu_table[order_id]
+
+                    if order_number_id == "----------------------------" then
+                        -- it's a separator
+                        self.menu_table[order_id][order_number] = self.separator
+                    elseif item_table[order_number_id] ~= nil then
+                        item_table[order_number_id].id = order_number_id
+                        self.menu_table[order_id][order_number] = item_table[order_number_id]
+                        item_table[order_number_id] = nil
+                    end
+                end
+
+            end
+        else
+            DEBUG("menu id not found:", order_id)
+        end
+    end
+    --attach orphans based on menu_hint
+    
+    -- now do the submenus
+    DEBUG("SUBMENUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUS")
+    DEBUG("self.sub_menus", self.sub_menus)
+    for i,sub_menu in ipairs(self.sub_menus) do
+        self.sub_menu_position = {}
+        self:findById(self.menu_table["KOMenu:menu_buttons"], sub_menu)
+        if self.sub_menu_position and self.sub_menu_position.id then
+            self.sub_menu_position.sub_item_table = self.menu_table[sub_menu]
+            self.menu_table[sub_menu] = nil
+        end
+    end
+    -- @TODO avoid this extra mini-loop
+    -- cleanup, top-level items shouldn't have sub_item_table
+    for i,top_menu in ipairs(self.menu_table["KOMenu:menu_buttons"]) do
+        self.menu_table["KOMenu:menu_buttons"][i] = self.menu_table["KOMenu:menu_buttons"][i].sub_item_table
+    end
+    
+    
+end
+
+function MenuSorter:findById(tbl, needle_id, result)
+
+
+--DEBUG("TBL given",tbl)
+    for k,v in pairs(tbl) do
+        if #self.sub_menu_position == 1 then
+            break
+        end
+        --DEBUG("FINDBYID:", needle_id, "current:", k,v)
+
+        if type(k) == "number" or k == "sub_item_table" then
+            if v.id == needle_id then
+                DEBUG("FOUND IT FOUND IT FOUND IT FOUND IT FOUND IT FOUND IT FOUND IT FOUND IT FOUND IT FOUND IT ", v.id)
+                self.sub_menu_position = v
+                break
+            elseif type(v) == "table" and v.id then
+                DEBUG("GOING DEEPER", v.id)
+                self:findById(v, needle_id)
+            end
+        end
+    end
```

Comment:
Haven't read your whole code, and I may be off topic, but just sharing something i discovered after hours of pulling hairs : ipairs stops iteration at first nil:
```
> a={"abd", "def", nil, "ghi"}
> for n, i in ipairs(a) do print(n, i) end
1       abd
2       def
> for n, i in pairs(a) do print(n, i) end
1       abd
2       def
4       ghi
```
Dunno if that's in play there, but sharing that just in case it is.